app/controllers/openid_connect/authorization_controller.rb

```diff
@@ -84,10 +105,26 @@ def session_params
       user_session.delete(:openid_auth_request) || {}
     end
 
+    def store_request
+      return if sp_session[:request_id]
+
+      @request_id = SecureRandom.uuid
+      ServiceProviderRequest.find_or_create_by(uuid: @request_id) do |sp_request|
+        sp_request.issuer = @authorize_form.client_id
+        sp_request.loa = @authorize_form.acr_values.sort.max
+        sp_request.url = request.original_url
+      end
+    end
+
     def add_sp_metadata_to_session
-      session[:sp] = { loa3: @authorize_form.loa3_requested?,
-                       issuer: @authorize_form.client_id,
-                       show_start_page: true }
+      return if sp_session[:request_id]
+
+      session[:sp] = {
```

Comment:
Good question. For several reasons.

- We need to store the request URL to make sure that existing users who sign in are redirected back to the SP. It's easier and faster to query the session rather than the DB, and easier than making sure the request_id is available on all forms and URLs throughout the auth process.
- The branded nav bar depends on the presence of either the issuer in the session or the request_id in the params. In the scenario where a user is signing in, we don't include the request_id params throughout the sign in process because that's only necessary for scenarios where users are switching browsers. Therefore, we need the issuer in the session.
- We have several places that check for `sp_session[:loa3]` in order to do stuff. In order to replace that, we'd need to pass in the `request_id` on every form and URL, and it would also mean a lot more DB lookups.